### PR TITLE
crop : sanitize values in crop and c&r

### PR DESCRIPTION
**Pull request by AlicVB**
this ensure more aggressively to have insane cropping values.
this will avoid dt to crash in case like #8491 or #8735 but doesn't solve the root cause...

### CODE REVIEW

**Comment by jenshannoschwalm** [comment]
As we dont understand the root cause of this, wouldn't it be better to make the clamping obvious to the user (by a popup or a module warning with tooltip) asking for opening an issue so we might understand?

**Comment by AlicVB** [comment]
hmm.. yes, it could be great. I have to think where to put that. Module warning may be the better if we want that average user notify us of the issue. I guess that I will have to learn how that works ;)

That said, this PR is needed anyway. I can come with a PR for the warning later ;)

**Comment by jenshannoschwalm** [comment]
There is a very simple example in the raw ca correction module.

**Comment by AlicVB** [comment]
@TurboGit : wait a minute before merging, I've spotted some other places in clipping which needs to be sanitize...

**Comment by AlicVB** [comment]
@TurboGit : that's ok for me now :)

@jenshannoschwalm : I've added the alert inside the iop. Thanks for the example !

File: src/iop/clipping.c [hunk 1]
```diff
@@ -451,6 +451,30 @@ static inline void transform(float *x, float *o, const float *m, const float t_h
...
+  dt_iop_clipping_params_t *p = (dt_iop_clipping_params_t *)self->params;
+  if(!g) return;
+
+  ++darktable.gui->reset;
+  if(g->clip_x != p->cx || g->clip_y != p->cy || g->clip_w != fabsf(p->cw) - p->cx
+     || g->clip_h != fabsf(p->ch) - p->cy)
+  {
+    gchar *er = dt_util_dstrcat(NULL,
+                                _("incorrect crop data have been detected\n"
+                                  "we have sanitize them, but this is a bug.\n"
```
Comment (jenshannoschwalm) [inline src/iop/clipping.c:466]: sanitized

**Comment by jenshannoschwalm** [comment]
The checks while in gui mode are good for me. Unfortunately - what about such checks also in `commit_params` to have this available in command line mode or generate-thumbnails?

**Comment by AlicVB** [comment]
@jenshannoschwalm : I've done the msg in commit_params too. That add again some messages on console in case of error, but well... it's an error so spamming the user about it shouldn't really be a problem ;)

**Comment by AlicVB** [comment]
for the record, that finally seems to fix #8491 for the c&r part.

**Comment by TurboGit** [comment]
This looks ok but not a bug fix so I've planed for 3.8.

**Comment by jenshannoschwalm** [comment]
Not for 3.6? I would appreciate to reconcider this because there have been at least two absolutely not understood issues and this pr would give a) dt not doing stupid things and b) we would get maybe relevant user feedback.

**Comment by TurboGit** [comment]
@jenshannoschwalm : I understand but what about possible side effects ? We've seen some "clear" changes that have introduced issues we had not envisioned. We are in bug fix only at this stage, and this is not bug fix so I would really prefer having this in 3.8. We have been living with the c&r issues since long time, we may still survive some more months.

As usual, this is only my point of view and we can always reconsider if more voices push for 3.6.

**Comment by AlicVB** [comment]
@TurboGit @jenshannoschwalm : What I can propose is to split this in 2 : 
1. only crop parameters sanitizing in commit_params, to avoid the crash. This can be considered as a bug fix with no real side effects, as anyway, null crop size is impossible to handle. We can even just ensure of non-zero sizes, putting apart the other cases handle by clampf... Note that this won't fix the gui part : in this specific case (not really common) user will have to reset c&r. But at least we won't crash !
2. all the rest of the check (gui, alerts, ...) that can be considered as an enhancement (or a not critical bugfix)

Of course, I perfectly know that c&r is... delicate ;) So as you say @TurboGit, we can also consider that it's a very uncommon case and postpone the fix...

**Comment by TurboGit** [comment]
@AlicVB : I agree that your proposal is safe and a good compromise. Let's go with that. TIA.

**Comment by AlicVB** [comment]
OK, the "critical" part is done is #9153 
I'll rebase this PR once it's merged (if it's ok for you)

**Comment by AlicVB** [comment]
this PR is now rebased on master (which include the critical part of the fix) so ready to be merged once 3.6 is released

**Review by TurboGit — CHANGES_REQUESTED** [review]
There is something wrong going on. I open a picture, reset crop&rotate params, I get free hand by default change size to get maybe half of picture move the area on the center and double click. I always get a red alert saying that there is something wrong with crop.

image

**Comment by AlicVB** [comment]
@TurboGit : Yes, I reproduce the issue. It's a "false positive" due to some precision error : we check that the params values are the same as the gui ones, but sometime (often) they differs a very little bit (I've not managed to know why)

We have 2 solutions : 
1. relax a bit the tests with something like `fabsf(g->clipx-p->cx)<0.0001`
2. remove the alert message

I would vote for 2, as 1 is a weird hack (how to define the precision is one example) and anyway c&r will be deprecated asap

**Comment by AlicVB** [comment]
@TurboGit : I've finally removed the alert, as explained in my previous post...

**Review by TurboGit — APPROVED** [review]
Looks good, thanks !

**Pull request merged**